libsignal-protocol/src/crypto.rs

```diff
@@ -1,23 +1,291 @@
+                let nid = openssl::nid::Nid::AES_192_CTR;
+                Cipher::from_nid(nid)
+                    .expect("OpenSSL should have AES_192_CTR !!")
+            },
+            (AesCtrNoPadding, 32) => Cipher::aes_256_ctr(),
+            (AesCbcPkcs5, 16) => Cipher::aes_128_cbc(),
+            (AesCbcPkcs5, 24) => {
+                let nid = openssl::nid::Nid::AES_192_CBC;
+                Cipher::from_nid(nid)
+                    .expect("OpenSSL should have AES_192_CBC !!")
+            },
+            (AesCbcPkcs5, 32) => Cipher::aes_256_cbc(),
+            _ => unreachable!(),
+        };
+        let block_size = signal_cipher_type.block_size();
+        let mut result = Vec::with_capacity(data.len() + block_size);
+        let mut crypter = Crypter::new(signal_cipher_type, mode, key, Some(iv))
+            .map_err(|_e| InternalError::Unknown)?;
+        crypter
+            .update(data, &mut result)
+            .map_err(|_e| InternalError::Unknown)?;
+        crypter
+            .finalize(&mut result)
+            .map_err(|_e| InternalError::Unknown)?;
+        Ok(result)
+    }
+}
+
+#[cfg(feature = "crypto-openssl")]
+impl Crypto for OpenSSLCrypto {
+    fn fill_random(&self, buffer: &mut [u8]) -> Result<(), InternalError> {
+        openssl::rand::rand_bytes(buffer).map_err(|_e| InternalError::Unknown)
+    }
+
+    fn hmac_sha256_init(&self, _key: &[u8]) -> Result<(), InternalError> {
+        let nid = openssl::nid::Nid::HMACWITHSHA256;
+        let ty = openssl::hash::MessageDigest::from_nid(nid)
+            .ok_or_else(|| InternalError::Unknown)?;
+        let ctx = openssl::hash::Hasher::new(ty)
+            .map_err(|_e| InternalError::Unknown)?;
+        let hmac_ctx = self.hmac_ctx.clone();
+        let mut guard = hmac_ctx.lock().map_err(|_e| InternalError::Unknown)?;
+        *guard = Some(ctx);
+        Ok(())
+    }
+
+    fn hmac_sha256_update(&self, data: &[u8]) -> Result<(), InternalError> {
+        let hmac_ctx = self.hmac_ctx.clone();
+        let mut guard = hmac_ctx.lock().map_err(|_e| InternalError::Unknown)?;
+        if let Some(ref mut ctx) = *guard {
+            ctx.update(data).map_err(|_e| InternalError::Unknown)?;
+        }
+        Ok(())
+    }
+
+    fn hmac_sha256_final(&self) -> Result<Vec<u8>, InternalError> {
+        let hmac_ctx = self.hmac_ctx.clone();
+        let mut guard = hmac_ctx.lock().map_err(|_e| InternalError::Unknown)?;
+        if let Some(ref mut ctx) = *guard {
+            ctx.finish()
+                .map(|buf| buf.as_ref().to_owned())
+                .map_err(|_e| InternalError::Unknown)
+        } else {
+            Err(InternalError::Unknown)
+        }
+    }
+
+    fn sha512_digest_init(&self) -> Result<(), InternalError> {
+        let ty = openssl::hash::MessageDigest::sha512();
+        let ctx = openssl::hash::Hasher::new(ty)
+            .map_err(|_e| InternalError::Unknown)?;
+        let sha512_ctx = self.sha512_ctx.clone();
```

Comment:
This `clone()` shouldn't be necessary. You can just do `let mut guard = self.sha512_ctx.lock() ...` directly.